### PR TITLE
[ONNXRuntime] Replaced CUDA binaries with source build for x86_64-linux-gnu

### PR DESCRIPTION
**Pull request by stemann**
Replaced CUDA binaries with source build for x86_64-linux-gnu

Also,
* Using CUDA platform selection (CUDA platform augmentation code block).
* Updated recipe wrt. Torch recipe (in particular wrt. CUDA approach)
* Reduced depth of git submodule update to 1.

Edit, 2025-01-29: Split recipe separate recipes ONNXRuntime, and ONNXRuntime_CUDA

### CODE REVIEW

File: O/ONNXRuntime/build_tarballs.jl [hunk 1]
```diff
@@ -122,11 +135,12 @@ dependencies = [
         compat = tensorrt_compat,
         platforms = cuda_platforms),
     Dependency("Zlib_jll"; platforms = cuda_platforms),
-    HostBuildDependency(PackageSpec("protoc_jll", Base.UUID("c7845625-083e-5bbe-8504-b32d602b7110"), v"3.16.1"))
+    BuildDependency(PackageSpec("CUDA_full_jll", first(cuda_versions)); platforms = cuda_platforms),
```
Comment (maleadt) [inline O/ONNXRuntime/build_tarballs.jl:138]: CUDA_full is deprecated and should be replaced by CUDA_SDK (and CUDA_SDK_static if needed).

Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:138]: I would like to (at first, in this PR - and in #4386) build this old version of ONNXRuntime, v1.10.0, to get support for older Nvidia Jetson devices ("JetPack 4"), e.g. Jetson Nano - which requires using CUDA 10.2 (and CUDNN v8.2 and [TensorRT v8.0](https://github.com/JuliaBinaryWrappers/TensorRT_jll.jl/releases/tag/TensorRT-v8.0.1%2B0)). Also, TensorRT v8.0 was only available for CUDA 11.3, and not 11.4, so for CUDA 11 I was forced to use CUDA 11.3 and not 11.4. It seems I then have to use CUDA_full and not CUDA_SDK, right?

What I am aiming for is getting as close as possible to the build versions listed in the tables on https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements and https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#requirements - while being able to use the binaries provided for the Jetson (aarch64) devices listed on https://elinux.org/Jetson_Zoo#ONNX_Runtime (i.e. v1.10 and v1.11) (and the binaries for Windows from e.g. https://github.com/microsoft/onnxruntime/releases/tag/v1.10.0)

Comment (maleadt) [reply O/ONNXRuntime/build_tarballs.jl:138]: Yeah CUDA_SDK only supports 11.4 and upwards.

Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:138]: How to best support this CUDA 10.2 scenario? I know that CUDA 10.2 support was dropped for CUDA_Runtime >= v0.3 (and CUDA.jl >= v4.1)...

I have also noticed that CUDA platform tag changed from including minor versions to specifying e.g. `cuda=11.0` or `cuda=12.0` for later versions...?

**Comment by stemann** [comment]
@maleadt Any updated comments to your remarks in #4386 ?
1. https://github.com/JuliaPackaging/Yggdrasil/pull/4386#discussion_r867716025
2. https://github.com/JuliaPackaging/Yggdrasil/pull/4386#pullrequestreview-1100536230
3. https://github.com/JuliaPackaging/Yggdrasil/pull/4386#discussion_r968181306

**Comment by stemann** [comment]
I see. Would it make sense to re-build TensorRT v8.0.1 (and v8.0.3) using the new approach? (That might entail also re-building CUDNN v8.2.1 and v8.2.4)

The TensorRT v8.0.1 artifacts for CUDA 11.0-11.3 was essentially just re-packaging of the same CUDA (11.3) binaries.

The approach would have to be adapted to also handle CUDA 10.2...

**Comment by stemann** [comment]
But for the latest CUDNN releases, the artifacts are built with e.g. `cuda="11.0"`, and not minor-version-specific platform specification...

File: O/ONNXRuntime/build_tarballs.jl (hunk 1 above)
Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:138]: ... so for ONNXRuntime v1.10-v1.11 (the versions supporting JetPack 4.6, i.e. CUDA 10.2), we have to stick with CUDA_full, right?

And then from ONNXRuntime v1.12, replace CUDA_full with with CUDA_SDK (and perhaps CUDA_SDK_static as a BuildDependency, to have the static cudart for CMake/CUDA configuration).

**Comment by maleadt** [comment]
> [CUDNN] artifacts are built with e.g. `cuda="11.0"`, and not minor-version-specific platform specification

Why isn't that OK? CUDNN binaries are only provided for 11.x and 12.x, so they are compatible across the entire release. Our platform augmentation uses semver-style matching, so should pick up the correct binary.

When building new artifacts, the new `CUDA.supported_platforms` will generate a build for every minor version, because the actual compatibility can vary on a case-by-case basis (binaries built for CUDA 11.0 aren't guaranteed to be compatible on all of 11.x, even though it's often the case). But that doesn't require the same of all dependencies, e.g., ONNX with a `cuda=11.2` artifact should work just fine with CUDNN `cuda=11.0`.

**Comment by stemann** [comment]
It might be just fine... I just saw that the more recent CUDNN builds had only one artifact per CUDA-platform-major version, so I was wondering whether the re-packaging of the TensorRT 11.3-binary should be re-done - to get a single artifact rather than four artifacts - which are all just copies of the same 11.3-binary.

I did also see that the Yggdrasil CUDA platform augmentation, and that there were some semver-based matching in [platforms/cuda.jl#L51](https://github.com/JuliaPackaging/Yggdrasil/blob/master/platforms/cuda.jl#L51), but the code/logic was a little hard to follow (probably due to the logic of `Base. BinaryPlatforms` comparison strategies)...

**Comment by stemann** [comment]
> > I was wondering whether the re-packaging of the TensorRT 11.3-binary should be re-done - to get a single artifact rather than four artifacts - which are all just copies of the same 11.3-binary
> 
> If it's just copies, then not. If it's a build, it might be safest to do so, unless you've verified that the binaries are forwards compatible.

I'm not completely sure what you are saying here... can you clarify?

The CUDA 11.0-11.3 artifacts for the TensorRT v8.0.1 JLL are just copies of the same CUDA 11.3-binary, cf. https://github.com/JuliaPackaging/Yggdrasil/pull/4347/files#diff-0d6f0803279fc06a9c714fb026ee87dd7ab9150c559ee167e63b2fd34e3f5af1

**Comment by maleadt** [comment]
I meant that if you're just copying the same files and tagging them with different CUDA tags, you shouldn't, and it's fine to only attach the `11.3` tag. Only if you actually build object files from sources using `nvcc` and the CUDA SDK, it may be safest to build for each CUDA toolkit version. But you're not doing that; I was just adding some context.

**Comment by stemann** [comment]
Right, then in that case the TensorRT v8.0 JLL's are a little bogus, I believe...

**Comment by stemann** [comment]
Seems like this one will be easier to complete once CUDA 10.2 has been restored, cf. #7623.

**Comment by stemann** [comment]
This is now using CUDA_SDK_jll instead of CUDA_full_jll, and thus CUDA 11.4. But TensorRT_jll v8.0.1 was built specifying CUDA 11.0-11.3, so there is no artifact for CUDA 11.4.

@maleadt Shouldn't rebuilding TensorRT_jll v8.0.1 with the more generic "v11.0" rather than the specific v11.0-v11.3 fix this?

**Comment by maleadt** [comment]
> But TensorRT_jll v8.0.1 was built specifying CUDA 11.0-11.3, so there is no artifact for CUDA 11.4.

That doesn't matter, and we will use 11.3's artifacts for 11.4: https://github.com/JuliaPackaging/Yggdrasil/blob/37f6397d2c6d566b129de337bb8f737f7914446c/platforms/cuda.jl#L50-L54

```julia
shell> cat Project.toml
[deps]
CUDA_Runtime_jll = "76a88914-d11a-5bdc-97e0-2f5a05c973a2"
TensorRT_jll = "2eaff018-c6f9-5ddb-97b6-8b0d18888fbb"

[preferences.CUDA_Runtime_jll]
version = "11.4"

julia> using CUDA_Runtime_jll, TensorRT_jll

julia> CUDA_Runtime_jll.host_platform
Linux x86_64 {cuda=11.4, cuda_local=false, cxxstring_abi=cxx11, julia_version=1.10.0, libc=glibc, libgfortran_version=5.0.0, libstdcxx_version=3.4.30}

julia> TensorRT_jll.is_available()
true

julia> TensorRT_jll.best_wrapper
"/home/tim/.julia/packages/TensorRT_jll/C392a/src/wrappers/x86_64-linux-gnu-cuda+11.3.jl"
```

Note that this binary-level compatibility is not guaranteed, is why it's recommended to build for every CUDA version.

> Shouldn't rebuilding TensorRT_jll v8.0.1 with the more generic "v11.0" rather than the specific v11.0-v11.3 fix this?

11.0 is not more generic.

**Comment by stemann** [comment]
OK.

I'm a bit puzzled that the build for CUDA 11.4 is failing - complaining that it cannot find TensorRT:
```
Warning: Dependency TensorRT_jll does not have a mapping for artifact TensorRT for platform x86_64-linux-gnu-libgfortran5-cxx11-cuda+11.4
```
On Monday: https://buildkite.com/julialang/yggdrasil/builds/7218#018c7e4c-7bc9-4cb8-84c8-3f021111fd52/619-721

... and what is even more puzzling, is that since we merged CUDNN v8.2.4 rebuild, it is now also complaining about CUDNN...
```
┌ Warning: Dependency TensorRT_jll does not have a mapping for artifact TensorRT for platform x86_64-linux-gnu-libgfortran5-cxx11-cuda+11.4
┌ Warning: Dependency CUDNN_jll does not have a mapping for artifact CUDNN for platform x86_64-linux-gnu-libgfortran5-cxx11-cuda+11.4
```
Wednesday (today): https://buildkite.com/julialang/yggdrasil/builds/7257#018c87d5-6c3e-4fab-b959-19b5c1b4cd89/620-732

**Comment by stemann** [comment]
@maleadt So, I think I finally figured this out :-)

It's pretty obvious by now, but anyway: CUDA platform tag must match when using dependencies in BinaryBuilder: E.g. if building, e.g., ONNXRuntime, for CUDA **11.4**, and using CUDNN as a dependency, the CUDNN _build version_ used must have a CUDA **11.4** artifact. And similarly for TensorRT and so on, down the dependency chain.

So I guess, given that it is often the case that binaries built for CUDA 11 are compatible across all CUDA 11 versions (and similar for CUDA 12, I assume), it should be preferable (as long as the binary compatibility holds), to only provide a single CUDA 11-platform artifact, e.g. for CUDA 11.0.

This would mean that e.g. TensorRT v8.0.1 or v8.0.3 for which Nvidia provides a CUDA 11.3 artifact, should be packaged as a CUDA 11.0 artifact (and no 11.1-11.3 copies of this). This sounds a little dodgy, but I hope it's true :-)

This would mean that this package, ONNXRuntime, should target CUDA 11.0 (and use CUDNN + TensorRT for CUDA 11.0), but use the CUDA SDK 11.4 when building.

Right? :-)

**Comment by maleadt** [comment]
> This would mean that e.g. TensorRT v8.0.1 or v8.0.3 for which Nvidia provides a CUDA 11.3 artifact, should be packaged as a CUDA 11.0 artifact (and no 11.1-11.3 copies of this).

I don't think this is correct. It generally holds that if you compile code for toolkit 11.0 it will work on 11.x (although there have been exceptions!), however, if you compile for a more recent version it may use features that are not available on earlier minor releases. So if TensorRT is distributed for CUDA 11.3, I wouldn't expect it to fully work on 11.2 or below.

**Comment by stemann** [comment]
Ah yes - that makes sense.

OK, then I guess the correct way is probably to build for CUDA 11.3 in this case (as this branch currently does) - as that is the version that has artifacts for the build versions of CUDNN and TensorRT.

**Comment by stemann** [comment]
@maleadt This should now also be up to current standards (for a combined CPU+CUDA JLL package) - similar to #9785.

File: O/ONNXRuntime/build_tarballs.jl [hunk 2]
```diff
@@ -90,23 +101,22 @@ function platform_exclude_filter(p::Platform)
...
-        push!(platforms, p)
-        push!(cuda_platforms, p)
+let cuda_platforms = CUDA.supported_platforms(min_version=v"10.2", max_version=v"11")
+    filter!(p -> !(arch(p) == "x86_64" && Sys.islinux(p) && p["cuda"] == "10.2"), cuda_platforms) # Fails with: nvcc error   : 'ptxas' died due to signal 11 (Invalid memory reference)
+    push!(cuda_platforms, Platform("x86_64", "Linux"; cuda = "11.3"))
+    push!(cuda_platforms, Platform("x86_64", "Windows"; cuda = "11.3"))
+
+     # Tag non-CUDA platforms matching CUDA platforms with cuda="none"
+    for platform in platforms
+        if CUDA.is_supported(platform) && arch(platform) != "powerpc64le" || platform == Platform("x86_64", "Windows")
```
Comment (maleadt) [inline O/ONNXRuntime/build_tarballs.jl:112]: Why those additional checks? That's supposed to be handled by `CUDA.is_supported`: https://github.com/JuliaPackaging/Yggdrasil/blob/6feb2ef0524c363cdacd2da97be459a0ee74e702/platforms/cuda.jl#L204

File: O/ONNXRuntime/build_tarballs.jl [hunk 3]
```diff
@@ -2,16 +2,19 @@
...
 version = v"1.10.0"
 
 # Cf. https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements
 # Cf. https://onnxruntime.ai/docs/execution-providers/TensorRT-ExecutionProvider.html#requirements
 cuda_versions = [
-    # No CUDA 10.2, since pre-built x86_64 CUDA binaries are built for CUDA 11
-    v"11.3", # Using 11.3, and not 11.4, to be compatible with TensorRT (JLL) v8.0.1 (which includes aarch64 support)
+    v"10.2",
+    v"11.3", # Using 11.3, and not 11.4, to be compatible with CUDNN v8.2.4, and with TensorRT (JLL) v8.0.1 (where the latter includes aarch64 support).
 ]
```
Comment (maleadt) [inline O/ONNXRuntime/build_tarballs.jl:17]: `cuda_versions` is not used anymore

File: O/ONNXRuntime/build_tarballs.jl [hunk 4]
```diff
@@ -90,23 +101,22 @@ function platform_exclude_filter(p::Platform)
...
-cuda_platforms = Platform[]
-for cuda_version in cuda_versions
-    cuda_tag = "$(cuda_version.major).$(cuda_version.minor)"
-    for p in [
-        Platform("x86_64", "Linux"; cuda = cuda_tag),
-        Platform("x86_64", "Windows"; cuda = cuda_tag)
-    ]
-        push!(platforms, p)
-        push!(cuda_platforms, p)
+let cuda_platforms = CUDA.supported_platforms(min_version=v"10.2", max_version=v"11")
```
Comment (maleadt) [inline O/ONNXRuntime/build_tarballs.jl:105]: Most people are on CUDA 12 now, so it'd be good to also bump the version and include all supported versions.

File: O/ONNXRuntime/build_tarballs.jl [hunk 5]
```diff
@@ -115,18 +125,44 @@ products = [
...
+for platform in platforms
+    should_build_platform(platform) || continue
+    additional_deps = BinaryBuilder.AbstractDependency[]
+    if haskey(platform, "cuda") && platform["cuda"] != "none"
+        if platform["cuda"] == "11.3"
+            additional_deps = BinaryBuilder.AbstractDependency[
+                BuildDependency(PackageSpec("CUDA_full_jll", v"11.3.1")),
+                Dependency("CUDA_Runtime_jll", v"0.7.0"), # Using v"0.7.0" to get support for cuda = "11.3" - using Dependency rather than RuntimeDependency to be sure to pass audit
+            ]
+        else
```
Comment (maleadt) [inline O/ONNXRuntime/build_tarballs.jl:141]: I'm not sure I like this, we pretty much decided we don't support 11.3 anymore. Why can't you bump TensorRT so that it supports a reasonable CUDA version?

File: O/ONNXRuntime/build_tarballs.jl (hunk 2 above)
Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:112]: I want to _exclude_ PowerPC64LE, and include Windows.

File: O/ONNXRuntime/build_tarballs.jl (hunk 4 above)
Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:105]: Yes - will do a modern version in a follow up.

File: O/ONNXRuntime/build_tarballs.jl (hunk 5 above)
Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:141]: … and to get changes done in ONNXRuntime.jl (which will also require an update to a modern version to get PRs through).

Comment (stemann) [reply O/ONNXRuntime/build_tarballs.jl:141]: … and verify that TensorRT works before pushing hundreds of MB to JuliaBinaryWrappers.

**Comment by stemann** [comment]
@maleadt I have separated the CUDA parts into a separate ONNXRuntime_CUDA recipe - not using the cuda+none stuff, but still using lazy artifacts for ONNXRuntime_CUDA.

Could you check if things are OK, and could you resolve those of your comments above that are "resolved" (i.e. postponed for future).

These recipes will need an update before being rolled into https://github.com/jw3126/ONNXRunTime.jl/pull/19, as the Julia project is now using v1.20+, but I would still like to merge this v1.10 first to get those parts done (including the jetson CUDA 10.2 support).

**Review by maleadt — APPROVED** [review]
I'll defer to you on this, assuming you verified it all works.

**Comment by stemann** [comment]
@giordano The register job failed for ONNXRuntime_CUDA - any idea what happened between my last commit and the merged commit?
* https://buildkite.com/julialang/yggdrasil/builds/16857
* https://buildkite.com/julialang/yggdrasil/builds/17076#0194cb5e-80a9-4fd8-abcc-2a31822cf69b/592-607

**Comment by giordano** [comment]
I tried to restart the job but it's still failing.  No clue what's the issue.  For future record:
```
ERROR: LoadError: Incomplete JLL release!  Could not find tarball for aarch64-linux-gnu-cxx03-cuda+10.2-cuda_platform+jetson
```

**Comment by stemann** [comment]
OK, thanks.

Yeah, I saw that.

Attempting a rebuild in #10426
